packages/filesystem/src/node/download/file-download-collection.ts

```diff
@@ -0,0 +1,89 @@
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+import { injectable, inject } from 'inversify';
+import { ILogger } from '@theia/core/lib/common/logger';
+import * as rimraf from 'rimraf';
+
+export interface DownloadStorageItem {
+    file: string;
+    root?: string;
+    size: number;
+    remove: boolean;
+    expire?: number;
+}
+
+@injectable()
+export class FileDownloadCollection {
+
+    @inject(ILogger)
+    protected readonly logger: ILogger;
+
+    protected readonly downloads = new Map<string, DownloadStorageItem>();
+    protected readonly expireTime: number = 1;
+
+    addDownload(id: string, downloadInfo: DownloadStorageItem): void {
+        downloadInfo.file = encodeURIComponent(downloadInfo.file);
+        if (downloadInfo.root) {
+            downloadInfo.root = encodeURIComponent(downloadInfo.root);
+        }
+        // expires in 1 minute enough for parallel connections to be connected.
+        downloadInfo.expire = new Date().getTime() + (this.expireTime * 60000);
+        this.downloads.set(id, downloadInfo);
+    }
+
+    getDownload(id: string): DownloadStorageItem | undefined {
+        this.expireDownloads();
+        const downloadInfo = this.downloads.get(id);
+        if (downloadInfo) {
+
+            downloadInfo.file = decodeURIComponent(downloadInfo.file);
+            if (downloadInfo.root) {
+                downloadInfo.root = decodeURIComponent(downloadInfo.root);
+            }
+        }
+        return downloadInfo;
+    }
+
+    deleteDownload(id: string): void {
+        const downloadInfo = this.downloads.get(id);
+        if (downloadInfo && downloadInfo.remove) {
+            this.deleteRecursively(downloadInfo.root || downloadInfo.file);
+        }
+        this.downloads.delete(id);
+    }
+
+    values(): { [key: string]: DownloadStorageItem } {
+        this.expireDownloads();
+        return [...this.downloads.entries()].reduce((downloads, [key, value]) => ({ ...downloads, [key]: value }), {});
+    }
+
+    protected deleteRecursively(pathToDelete: string): void {
+        rimraf(pathToDelete, error => {
+            if (error) {
+                this.logger.warn(`An error occurred while deleting the temporary data from the disk. Cannot clean up: ${pathToDelete}.`, error);
+            }
+        });
+    }
+
+    protected expireDownloads(): void {
```

Comment:
[question]: What happens, when the user performs `Copy Download Link` multiple times but does not eventually download anything? Wouldn't it be better to have a timer task that removes obsolete entries?

Reply:
I thought of that as well. But i'm just not sure if it is worth adding a timer here or if it will affect any performance.  If a user clicks on Copy Download Link multiple times, in less than 1 minute, those ids will get stored on cache and the corresponding files ont /tmp folder. And soon as the user clicks that Copy Download Link again, it will expire all the others. I get your point that what if a user just dont ever click on it again or if the size affects the whole disk performance etc.  

If you think a simple timer is fine here, then i can go for that.